File: 프로그래머스/1/250121. ［PCCE 기출문제］ 10번 ／ 데이터 분석/Tae Seong Hong/［PCCE 기출문제］ 10번 ／ 데이터 분석.cpp
#include <string>
#include <vector>

using namespace std;
int change_to_index(string calc_string) {
    if (calc_string == "code")
        return 0;
    else if (calc_string == "date")
        return 1;
    else if (calc_string == "maximum")
        return 2;
    else if (calc_string == "remain")
        return 3;
}
    
vector<vector<int>> solution(vector<vector<int>> data, string ext, int val_ext, string sort_by) {
    vector<vector<int>> answer;
    
    int ext_index = change_to_index(ext);
    int sort_index = change_to_index(sort_by);
    int max = 0;
    
    for(int i=0;i<data.size();i++) {
        if(data[i][ext_index] < val_ext) {
            answer.push_back(data[i]);
            for(int a=0;a<answer.size()-1;a++) {
                for(int b=0;b<answer.size()-1-a;b++) {
                    if(answer[b][sort_index] > answer[b + 1][sort_index]) {
                        vector<int> temp = answer[b];
                        answer[b] = answer[b + 1];
                        answer[b + 1] = temp;
                        
                    }
                }
            }
        }
    }
    return answer;
}

File: 프로그래머스/1/250121. ［PCCE 기출문제］ 10번 ／ 데이터 분석/Hyun-suck Lee/［PCCE 기출문제］ 10번 ／ 데이터 분석.java
import java.util.*;

class Solution {
    public int[][] solution(int[][] datas, String ext, int valExt, String sortBy) {
        // ext나 sortBy 등 명칭으로 인덱스 찾는 용도
        Map<String, Integer> fieldMap = new HashMap<>();
        
        String[] fields = {"code", "date", "maximum", "remain"};
        for (int idx = 0; idx < fields.length; idx++) {
            fieldMap.put(fields[idx], idx);
        }

        
        // 결과값 개수 모르니 List로 
        List<int[]> calc = new ArrayList<>();
        
        // data 에서 ext(field)값이 valExt보다 작은 데이터만 선택
        int baseExtIdx = fieldMap.get(ext);
        for (int[] data : datas) {
            if (data[baseExtIdx] >= valExt) continue;
            calc.add(data);
        }
        
        // sortBy 기준으로 오름차순
        int baseSortIdx = fieldMap.get(sortBy);
        calc.sort( (a1, a2) -> Integer.compare(a1[baseSortIdx], a2[baseSortIdx]) );
        
        int[][] result = new int[calc.size()][calc.get(0).length];
        for (int idx = 0; idx < calc.size(); idx++) {
            result[idx] = calc.get(idx);
        }
        
        return result;
    }
}